GSRs Unavailable

raunaqsingh2020 opened the pull request:
Changed GSRController to show closed view when there are 0 rooms and to show the empty view (no results found) only when there are rooms but none available for the specific times.

E.g. GSR with rooms: 
![Screen Shot 2020-10-07 at 6 12 21 PM](https://user-images.githubusercontent.com/38117531/95393554-d1b95e80-08c8-11eb-969b-856bed125559.png)

E.g. GSR with rooms but none available at time given by slider:
![Screen Shot 2020-10-07 at 6 13 17 PM](https://user-images.githubusercontent.com/38117531/95393662-f3b2e100-08c8-11eb-98f6-7f148510589e.png)

E.g. GSR with 0 rooms:
![Screen Shot 2020-10-07 at 6 12 32 PM](https://user-images.githubusercontent.com/38117531/95393677-fa415880-08c8-11eb-98a6-b9a1eb1641e0.png)

r3khaan commented on PennMobile/GSR-Booking/Views/ClosedView.swift:
```diff
@@ -0,0 +1,52 @@
+    }()
+    
+    private let closedImage: UIImageView = {
+        let imgView = UIImageView(image: UIImage(named: "EmptyStateGSR")!)
```
Not a big deal, but I don't think we need to force unwrap the `UIImage` here?

r3khaan commented on PennMobile/GSR-Booking/Views/ClosedView.swift:
```diff
@@ -0,0 +1,52 @@
+        
+        backgroundColor = .clear
+        
+        addSubview(closedImage)
+        closedImage.widthAnchor.constraint(equalToConstant: 200).isActive = true
+        closedImage.centerYAnchor.constraint(equalTo: centerYAnchor, constant: -100).isActive = true
+        closedImage.centerXAnchor.constraint(equalTo: centerXAnchor).isActive = true
+        
+        addSubview(label)
+        
+        label.centerYAnchor.constraint(equalTo: centerYAnchor, constant: 75).isActive = true
+        label.centerXAnchor.constraint(equalTo: centerXAnchor).isActive = true
+        label.widthAnchor.constraint(equalTo: widthAnchor, constant: -10).isActive = true
```
Nice!!

r3khaan commented on PennMobile/GSR-Booking/Views/ClosedView.swift:
```diff
@@ -0,0 +1,52 @@
+//
+//  ClosedView.swift
```
Let's call this file/class something more specific, like `GSRClosedView`

r3khaan commented on PennMobile/GSR-Booking/Controllers/GSRController.swift:
```diff
@@ -146,12 +157,16 @@ extension GSRController: GSRViewModelDelegate {
     func refreshDataUI() {
-        emptyView.isHidden = !viewModel.isEmpty
         tableView.isHidden = viewModel.isEmpty
+        closedView.isHidden = viewModel.existsNonEmptyRoom()
```
Ahh .. I totally see why `existsNonEmptyRoom` is not good enough for our purposes here. One way to resolve it:

Instead of checking whether at least one `room` exists, we should check if at least one `timeSlot` exists! If 0 `timeSlot`'s exists, then we should show the `closedView`. Otherwise, we should hide it.

jtlieb replied:
Just want to add something here -- I think we may want to add a different empty state rather than "closed" for when there are no empty rooms available. Def agree with Rehaan here, "Closed" should be used for an empty schedule for the building, which would signify that it's being closed for a significant amount of time"

r3khaan reviewed (changes requested):
Great start, Raunaq! I'm glad you were able to figure out the programmatic UI. I added some small comments here and there, but here are the major points:

Ideally, we want to show the same "closed" message + image for each GSR location (because all of them are actually closed right now). I should have explained that better, because I'm sure you're still a little confused about what a GSR location versus room versus timeslot is. 
- I added a suggestion for how to handle the case where a GSR location has >=1 rooms, but has 0 timeslots in the comments. 
- Try to see if you can figure out how to handle the case where the server doesn't return a valid response (e.g. for Huntsman). Hint: go to the `fetchData` function in the `GSRController.swift` file and see if you can figure it out.

jtlieb commented:
From a layout perspective, I think the illustration needs to be brought down closer to the center of the whitespace, currently it feels a little bit cramped below the slider bar.

raunaqsingh2020 commented:
Fixed the layout for the closed view by lowering the sign a bit. Fixed when this view shows:

When the app cannot get a valid response from the server (ex. Huntsman):

![Screen Shot 2020-10-11 at 7 31 25 PM](https://user-images.githubusercontent.com/38117531/95693648-68548b00-0bfb-11eb-85b8-a03bcf3e5139.png)

When the GSR has 0 `timeslots`:

![Screen Shot 2020-10-11 at 7 31 52 PM](https://user-images.githubusercontent.com/38117531/95693664-6f7b9900-0bfb-11eb-8d8a-a17b617f979a.png)

GSR with 1+ `timeslots` (added these manually to test):

![Screen Shot 2020-10-11 at 7 32 05 PM](https://user-images.githubusercontent.com/38117531/95693666-730f2000-0bfb-11eb-9437-7188130c0f5d.png)

GSR with 1+ `timeslots` when the slider does not include the times:

![Screen Shot 2020-10-11 at 7 32 14 PM](https://user-images.githubusercontent.com/38117531/95693667-74d8e380-0bfb-11eb-9f53-a9d90bffa018.png)

Let me know if there are any other changes to make. Thanks!

r3khaan commented:
@raunaqsingh2020 Hey, will take a look soon

r3khaan commented on PennMobile/GSR-Booking/ViewModel/GSRViewModel.swift:
```diff
@@ -202,6 +202,11 @@ extension GSRViewModel: GSRSelectionDelegate {
         currentSelection = []
     }
+    
+    func existsTimeSlot() -> Bool {
+        let roomsWithTimeSlots = allRooms.filter { $0.timeSlots.count > 0 }
+        return roomsWithTimeSlots.count > 0
+    }
```
this is some clean code 😎

r3khaan commented on PennMobile/GSR-Booking/Controllers/GSRController.swift:
```diff
@@ -134,24 +145,31 @@ extension GSRController: GSRViewModelDelegate {
         let date = viewModel.getSelectedDate()
         
         self.startLoadingViewAnimation()
-        
+                        
         GSRNetworkManager.instance.getAvailability(for: location.lid, date: date) { (rooms) in
             
             DispatchQueue.main.async {
-                if let rooms = rooms {
+                if let rooms = rooms {                    
                     self.viewModel.updateData(with: rooms)
                     self.refreshDataUI()
                     self.rangeSlider.reload()
                     self.refreshBarButton()
                     self.stopLoadingViewAnimation()
+                }else{//no valid response
+                    self.viewModel.updateData(with: [])
+                    self.refreshDataUI()
+                    self.rangeSlider.reload()
+                    self.refreshBarButton()
+                    self.stopLoadingViewAnimation()
```
Suggestion for readability: You can remove the `if ... else ...` and instead write `let rooms = rooms ?? []` at the start. This removes the duplicate code.

r3khaan reviewed (changes requested):
Added a minor formatting suggestion for code readability. Thanks for providing screenshots of the different cases, and testing for when there actually are timeslots.

r3khaan reviewed (approved):
LGTM!
